KEP-3015 PreferSameZone/PreferSameNode traffic distribution

danwinship opened the pull request:
#### What this PR does / why we need it:
Implements the new PreferSameNode traffic distribution mode, and the new PreferSameZone alias for PreferClose.

This doesn't update any documentation around PreferClose yet since PreferSameZone is still alpha.

#### Special notes for your reviewer:
integration and e2e tests TBD

#### Does this PR introduce a user-facing change?
```release-note
When the `PreferSameTrafficDistribution` feature gate is enabled, a new
`trafficDistribution` value `PreferSameNode` is available, which attempts to
always route Service connections to an endpoint on the same node as
the client. Additionally, `PreferSameZone` is introduced as an alias for
`PreferClose`.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/3015
```

/kind feature
/assign @thockin @aojea @gauravkghildiyal 
/priority important-soon
/triage accepted

danwinship committed:
Add PreferSameTrafficDistribution feature gate and associated API.

danwinship committed:
update generate docs for PreferSameTrafficDistribution

danwinship committed:
Implement the EndpointSlice controller side of PreferSameZone/PreferSameNode

danwinship committed:
Implement PreferSameNode traffic distribution in kube-proxy

danwinship committed:
Minor updates to kube-proxy topology code

Clarify the comments around terminating endpoints.

Remove stale references to the ProxyTerminatingEndpoints feature gate
in the unit tests.

danwinship committed:
Clarify hostname vs node name in kube-proxy

Various parts of kube-proxy passed around a "hostname", but it is
actually the name of the *node* kube-proxy is running on, which is not
100% guaranteed to be exactly the same as the hostname. Rename it
everywhere to make it clearer that (a) it is definitely safe to use
that name to refer to the Node, (b) it is not necessarily safe to use
that name with DNS, etc.

danwinship committed:
Minor updates to traffic distribution unit tests

Merge TestReconcileHints_trafficDistribution_is_PreferClose and
TestReconcileHints_trafficDistribution_is_nil_or_empty together.
Change the `trafficDistribution: ""` test to `trafficDistribution:
Unknown`, since `""` is not actually a possible value (but we should
still test that unknown values are ignored, to prevent weird skew
bugs).

Fill in the NodeName field in the endpoints. It's not needed yet but
it will be.

danwinship committed:
Don't write out dummy zone hints in clusters with no zones

If you set `trafficDistribution: PreferClose` on a service in a
cluster with no defined zones, then it would add

  hints:
    forZones:
    - name: ""

to each endpoint. This ended up working anyway since kube-proxy would
likewise end up looking for an endpoint for the "" zone, but it's
unnecessary, since you'd get exactly the same behavior by just leaving
all of the endpoints unhinted. (Of course there's no point in using
PreferClose traffic distribution in this case, but this will make
PreferSameNode cleaner.)

danwinship commented on pkg/features/kube_features.go:
```diff
@@ -291,6 +291,12 @@ const (
+	// kep: https://kep.k8s.io/3015
+	//
+	// Enables PreferSameZone and PreferSameNode values for trafficDistribution
+	ImprovedTrafficDistribution featuregate.Feature = "ImprovedTrafficDistribution"
```
This is the feature gate name I put in the KEP but having actually written all the code now, it sounds like it means "makes the distribution of traffic better", not "makes TrafficDistribution better", so I could rename it...

adrianmoisey replied:
It does sound like the distribution of traffic is better.
May be `ExpandedTrafficDistribution` or `ExtendedTrafficDistribution` is better, but that doesn't leave much for a next step.

`PerZoneAndPerNodeTrafficDistribution` is a long.

gauravkghildiyal replied:
`ExpandedTrafficDistribution` sounds good to me too. A future option can be something more specific -- the reason why we can't be more specific is because we're trying to account for both PreferSameNode and PreferSameZone right.

thockin replied:
I think "PreferSameTopologyTrafficDistribution" or "PreferSameTrafficDistribution" would work.  "Improved" is a little misleading.  Expanded is not much better.  But this doesn't matter a whole lot.

adrianmoisey commented on staging/src/k8s.io/endpointslice/reconciler.go:
```diff
@@ -51,9 +51,13 @@ type Reconciler struct {
+	// improvedTrafficDistribution determines if the new (PreferSameZone /
+	// PreferSameNode) trafficDistribution values should be considered when
+	// reconciling EndpointSlice hints.
+	improvedTrafficDistribution bool
```
I don't like this suggestion, but I'll make it and see what you think:
```suggestion
	improvedTrafficDistributionEnabled bool
```

Mostly to match `trafficDistributionEnabled`

It makes it pretty long, which is annoying.
This is also not some externally facing API, so I guess the shorter name is fine.

adrianmoisey commented on pkg/proxy/topology.go:
```diff
@@ -140,48 +144,77 @@ func CategorizeEndpoints(endpoints []Endpoint, svcInfo ServicePort, nodeLabels m
-		if !foundZone || zone == "" {
-			klog.V(2).InfoS("Skipping topology aware endpoint filtering since node is missing label", "label", v1.LabelTopologyZone)
-			return false
+		if endpoint.NodeHints().Len() == 0 {
+			allEndpointsHaveNodeHints = false
+		} else if endpoint.NodeHints().Has(nodeName) {
+			hasEndpointForNode = true
```
Another nit/opinion:
Could this logic move above the ZoneHint logic?

In my head the order is always node, zone, everything else, and it may make the code more readable

danwinship replied:
makes sense

gauravkghildiyal commented on staging/src/k8s.io/endpointslice/trafficdist/trafficdist.go:
```diff
@@ -101,43 +109,61 @@ func (defaultHeuristic) update(slice *discoveryv1.EndpointSlice) {
 		}
 
-		if endpoint.Hints == nil || len(endpoint.Hints.ForZones) != 1 || endpoint.Hints.ForZones[0].Name != zone {
+		wantNodeHint := endpoint.NodeName != nil && h.generateNodeHints
```
I'm wondering about the case where generateNodeHInts=false (maybe someone switch trafficDist field) and the nodeHint is currently incorrect. In such a case, the current behaviour seems to not remove the hint. I know our own code will never lead to such a state (unless there's a bug somewhere). Wondering if that's worth accounting for or not?

danwinship replied:
the code is supposed to deal with all cases, but you're right, it doesn't

gauravkghildiyal commented on staging/src/k8s.io/api/discovery/v1/types.go:
```diff
@@ -159,10 +159,17 @@ type EndpointConditions struct {
+	// This is an Alpha feature and is only used when the ImprovedTrafficDistribution
+	// feature gate is enabled.
+	// +listType=atomic
+	ForNodes []string `json:"forNodes,omitempty" protobuf:"bytes,2,name=forNodes"`
```
I know you'd have already given this some thought, but wanted us to acknowledge this in the PR too. The question is whether or not to be consistent with the `ForZones` data type. 

The original API design used a dedicated type citing possible future expansion https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2433-topology-aware-hints#future-api-expansion. 

I can imagine we may never need this for the nodes, but was wondering if it still has some value to be consistent and define a `ForNodes` type?

danwinship replied:
Discussed in the KEP PR [here](https://github.com/kubernetes/enhancements/pull/5140#discussion_r1947996701):

> The `ForZone` struct was apparently added to support future ideas like having multiple zones with `weight` values...
> 
> I guess we could do the same for `ForNodes`... ? I think the idea of weighted hints is based more on the "smart controllers / ambiguous API" model of TopologyAwareHints, while TrafficDistribution has moved more toward "simple controllers / explicit API" so that sort of future API expansion seems unlikely.

OTOH, I did repeatedly find it slightly weird when writing the code that ForNodes didn't behave the same way as ForZones, so maybe I should just make it consistent even if we don't have any use cases for the struct...

gauravkghildiyal replied:
Thanks for sharing the previous discussion link, sorry missed that conversation.

> OTOH, I did repeatedly find it slightly weird when writing the code that ForNodes didn't behave the same way as ForZones, so maybe I should just make it consistent even if we don't have any use cases for the struct...

Yep, while going through the code is when I too started wondering if having that consistency was a net positive. Seeing you share the same thoughts, I'm in favour of transitioning to a struct.

gauravkghildiyal reviewed (commented):
Thanks @danwinship. Besides a few nitpicks, this mostly looks good to me.

thockin commented on pkg/apis/discovery/types.go:
```diff
@@ -133,9 +133,15 @@ type EndpointConditions struct {
+	// using topology aware routing.
+	// This is an Alpha feature and is only used when the ImprovedTrafficDistribution
+	// feature gate is enabled. May contain a maximum of 8 entries.
+	ForNodes []ForNode
```
Do we have an upper bound on this array?  How many nodes are allowed? 16?  128? 1024?  We know that clusters are unlikely to have hundreds of zones, so it matters less (but should still be defined).  I think we don't yet have a use-case for more than 1 item here, though I don't mind leaving this open.

Alternately, it could be limited to 1 for now, and the fact that it is a slaice is "for future use".

Edit, I see 8 in the validation.  Do you want 8 or 1?

danwinship replied:
I copied 8 from the ForZones validation. The multiple zones is also "for future use", but we allow that, so I allowed it for nodes too. But I don't have a strong opinion.

(In theory, someone could implement "prefer same rack" by writing out a slice with a multi-element ForNodes hints by hand...)

gauravkghildiyal replied:
> Who is approving impl? I likely only have time for API in the next 3 days.

If that just means a second pair of eyes, I am happy to help. But if we do require a review from some actual OWNER, that I don't know.

thockin commented on pkg/apis/discovery/validation/validation.go:
```diff
@@ -221,7 +222,7 @@ func validateHints(endpointHints *discovery.EndpointHints, fldPath *field.Path)
 	}
 
-	zoneNames := sets.String{}
+	zoneNames := sets.Set[string]{}
```
Given how often EndpointSlices are updated, we should try to minimize wasteful allocations.  A set is a map which is several allocations under the covers.  Could I talk you into doing this as `make([]string, 0, len(endpointHints.ForZones))` and doing a linear search?  (with comment on why).  Or even a `[maxZoneHints]string`  and `nzones int`?

danwinship replied:
Done (with the slice version so I could use `slices.Contains`), and updated the port name validation as well. I left `supportedAddressTypes` and `supportedPortProtocols` using sets since those sets are just created once.

thockin commented on pkg/apis/discovery/validation/validation.go:
```diff
@@ -235,5 +236,25 @@ func validateHints(endpointHints *discovery.EndpointHints, fldPath *field.Path)
+		return allErrs
+	}
+
+	nodeNames := sets.Set[string]{}
```
ditto on set vs list

thockin commented on pkg/apis/discovery/validation/validation_test.go:
```diff
@@ -583,6 +584,71 @@ func TestValidateEndpointSlice(t *testing.T) {
+				}},
+			},
+		},
+		"too-many-node-hints": {
```
maybe a testcase for "" as a hint (or 2x "" to fail both parts)

thockin commented on pkg/registry/discovery/endpointslice/strategy.go:
```diff
@@ -149,26 +149,38 @@ func dropDisabledFieldsOnCreate(endpointSlice *discovery.EndpointSlice) {
 				endpointSlice.Endpoints[i].Hints = nil
```
The `if` line above this is redundant - can you clean it up while you are in here?

danwinship replied:
cleaned up and synced `dropDisabledFieldsOnCreate` and `dropDisabledFieldsOnUpdate`.

thockin reviewed (commented):
Focused on API bits

danwinship commented on pkg/apis/core/types.go:
```diff
@@ -4554,6 +4554,22 @@ const (
+	// preference will not result in endpoints getting overloaded.
+	// This is an Alpha feature and is only recognized if the
+	// ImprovedTrafficDistribution feature gate is enabled.
+	ServiceTrafficDistributionPreferSameZone = "PreferSameZone"
```
I consider it to be an alias, but I didn't update the docs to say so since it's still Alpha... I was planning to do that when we go to Beta.

danwinship committed:
Update the use of sets in EndpointSlice validation

Don't use sets for validating port name and zone hint uniqueness,
since constructing a new set each time is likely to be less efficient
than just doing a linear search.

Keep the sets for supportedAddressTypes and supportedPortProtocols
(since they're only constructed once) but switch to the generic set
API.

danwinship commented:
/retest

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6194,8 +6195,21 @@ func validateServiceTrafficDistribution(service *core.Service) field.ErrorList {
-	if *service.Spec.TrafficDistribution != v1.ServiceTrafficDistributionPreferClose {
-		allErrs = append(allErrs, field.NotSupported(field.NewPath("spec").Child("trafficDistribution"), *service.Spec.TrafficDistribution, []string{v1.ServiceTrafficDistributionPreferClose}))
+	var supportedTrafficDistribution []string
+	if !utilfeature.DefaultFeatureGate.Enabled(features.PreferSameTrafficDistribution) {
```
pedantically, this could be something like:

```
var supportedTrafficDistribution = func() sets.Set[string] {
    if !utilfeature.DefaultFeatureGate.Enabled(features.PreferSameTrafficDistribution) {
	    return sets.New(v1.ServiceTrafficDistributionPreferClose)
    }
	return sets.New(
			v1.ServiceTrafficDistributionPreferClose,
			v1.ServiceTrafficDistributionPreferSameZone,
			v1.ServiceTrafficDistributionPreferSameNode)
}()

func validateServiceTrafficDistribution(...) {
  ...
}
```

But service create/update isn't nearly as frequent as endpoints, so it matters a little less.

danwinship replied:
That wouldn't work though, because `--feature-gates` hasn't been processed yet at variable-initialization time, right? (And also it would break `SetFeatureGateDuringTest`.)

thockin replied:
oh, right, good point.  Would need an accessor function and a sync.Once.  Nevermind.

thockin reviewed (commented):
API bits are good

/approve for API

gauravkghildiyal commented:
/lgtm

(needs rebase)

thockin commented:
/lgtm

adrianmoisey commented on pkg/apis/discovery/validation/validation.go:
```diff
@@ -172,24 +174,26 @@ func validatePorts(endpointPorts []discovery.EndpointPort, fldPath *field.Path)
 	}
 
-	portNames := sets.String{}
+	// Even though a sets.Set would be more idiomatic, we use a []string here to avoid
+	// extra allocations (especially since there are presumably only a few ports anyway).
```
Another nit from me:
Could the reason behind this be added? Specifically the "Given how often EndpointSlices are updated" part from https://github.com/kubernetes/kubernetes/pull/130844#discussion_r2000274108

When I read the comment I was trying to figure out why we were trying to avoid extra allocations, the context of frequently updated EndpointSlices may be useful to someone that doesn't have that context

thockin replied:
That's fair.

I'm not going as far as demanding zero-allocation logic, but looking to
avoid obviously wasteful patterns in obviously frequent paths :)

thockin commented:
That's fair.

I'm not going as far as demanding zero-allocation logic, but looking to
avoid obviously wasteful patterns in obviously frequent paths :)


On Wed, Mar 19, 2025 at 12:36 PM Adrian Moisey ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In pkg/apis/discovery/validation/validation.go
> <https://github.com/kubernetes/kubernetes/pull/130844#discussion_r2004154320>
> :
>
> > +	// Even though a sets.Set would be more idiomatic, we use a []string here to avoid
> +	// extra allocations (especially since there are presumably only a few ports anyway).
>
> Another nit from me:
> Could the reason behind this be added? Specifically the "Given how often
> EndpointSlices are updated" part from #130844 (comment)
> <https://github.com/kubernetes/kubernetes/pull/130844#discussion_r2000274108>
>
> When I read the comment I was trying to figure out why we were trying to
> avoid extra allocations, the context of frequently updated EndpointSlices
> may be useful to someone that doesn't have that context
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/130844#pullrequestreview-2699938566>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKWAVAMPGIREA3UCWVQRQL2VHBNNAVCNFSM6AAAAABZC2PTKSVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMOJZHEZTQNJWGY>
> .
> You are receiving this because you were assigned.Message ID:
> <kubernetes/kubernetes/pull/130844/review/2699938566 <(269)%20993-8566>@
> github.com>
>
